finetrainers/utils/torch_utils.py

```diff
@@ -6,3 +9,21 @@ def unwrap_model(accelerator: Accelerator, model):
     model = accelerator.unwrap_model(model)
     model = model._orig_mod if is_compiled_module(model) else model
     return model
+
+
+def align_device_and_dtype(
+    x: Union[torch.Tensor, Dict[str, torch.Tensor]],
+    device: Optional[torch.device] = None,
+    dtype: Optional[torch.dtype] = None,
+):
+    if isinstance(x, torch.Tensor):
+        if device is not None:
+            x = x.to(device)
+        if dtype is not None:
+            x = x.to(dtype)
+    elif isinstance(x, dict):
+        if device is not None:
+            x = {k: align_device_and_dtype(v, device, dtype) for k, v in x.items()}
+        if dtype is not None:
+            x = {k: align_device_and_dtype(v, device, dtype) for k, v in x.items()}
+    return x
```

Comment:
I think this is an overkill given it's just iterating over a dict and performing `to()`. IMO, an in-line `to()` doesn't necessarily hurt even if it contributes to increased LoC.

Reply:
Not really, this will be hard to handle and create a mess of if-statements. We don't beforehand know all the kinds of prepared conditions by name. I mean, we do because we will be implementing it but there can be a lot of different names in the dictionaries + this would allow any custom implementations from users to directly benefit from just adding any more amount of tensors in the conditions they want and it would "just work". We can have dictionary keys as prompt_embeds, pooled_prompt_embeds, latents, control_latents, etc. to name a few, as well as single tensors and list of tensors nested in dictionaries, and many more as we move forward. Having an if-statement to check if a key exists and then move it to correct device would not be very nice